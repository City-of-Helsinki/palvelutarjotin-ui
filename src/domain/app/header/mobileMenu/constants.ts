export const MOBILE_MENU_ACTIONS = {
  CLOSE_MOBILE_MENU: 'CLOSE_MOBILE_MENU',
  OPEN_MOBILE_MENU: 'OPEN_MOBILE_MENU',
};
