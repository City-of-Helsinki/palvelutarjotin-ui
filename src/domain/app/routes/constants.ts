export enum ROUTES {
  EVENTS_SEARCH = '/search',
  EVENT_DETAILS = '/event/:id',
  CREATE_ENROLMENT = '/event/:id/enrolment/create',
  HOME = '/',
  PRIVACY_POLICY = '/privacy-policy',
  CMS_PAGE = '/cms-page/:slug',
  NEWSLETTER_SUBSCRIBE = '/newsletter/subscribe',
}

export enum PATHNAMES {
  CMS_PAGE = '/cms-page/[...slug]',
}
