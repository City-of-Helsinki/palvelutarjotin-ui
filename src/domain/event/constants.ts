export const OCCURRENCE_LIST_PAGE_SIZE = 10;

export const EVENT_PLACEHOLDER_IMAGES = [
  '/static/images/event_placeholder_A.jpg',
  '/static/images/event_placeholder_B.jpg',
  '/static/images/event_placeholder_C.jpg',
  '/static/images/event_placeholder_D.jpg',
];
