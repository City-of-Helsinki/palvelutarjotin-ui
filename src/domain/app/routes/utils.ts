import { CollectionItemType } from 'react-helsinki-headless-cms';

import { ROUTES } from './constants';

export const getCmsPagePath = (uri?: string | null): string => {
  return ROUTES.CMS_PAGE.replace('/:uri', uri ?? '');
};

export const getCmsArticlePath = (uri?: string | null): string => {
  return ROUTES.CMS_ARTICLE.replace('/:uri', uri ?? '');
};

export const getCollectionItemUrl = (item: CollectionItemType): string => {
  if (!item) {
    return '#';
  }
  if (item.__typename === 'Post') {
    return getCmsArticlePath(item.uri);
  }
  if (item.__typename === 'Page') {
    return getCmsPagePath(item.uri);
  }
  return item?.uri ?? '';
};
