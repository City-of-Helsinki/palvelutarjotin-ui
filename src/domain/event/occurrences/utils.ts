import {
  EventFieldsFragment,
  OccurrenceFieldsFragment,
} from '../../../generated/graphql';
import { ENROLMENT_ERRORS } from '../../enrolment/constants';
import {
  hasOccurrenceSpace,
  isEnrolmentClosed,
  isOccurrenceCancelled,
} from '../../occurrence/utils';

export function getEnrolmentError(
  occurrence: OccurrenceFieldsFragment,
  event: EventFieldsFragment
): ENROLMENT_ERRORS | null {
  if (isEnrolmentClosed(occurrence, event))
    return ENROLMENT_ERRORS.ENROLMENT_CLOSED_ERROR;
  if (!hasOccurrenceSpace(occurrence))
    return ENROLMENT_ERRORS.NOT_ENOUGH_CAPACITY_ERROR;
  if (isOccurrenceCancelled(occurrence)) {
    return ENROLMENT_ERRORS.ENROLMENT_CANCDELLED_ERROR;
  }

  return null;
}

export function getOrderedLanguages(
  languages: { id: string; name: string }[]
): string[] {
  const languageOrder = ['fi', 'sv', 'en'];
  const languageStrings = languages.map((l) => l.id.split('_')[0]);

  // order languages based on languageOrder, other languages go
  // to the end of the array
  return languageStrings.sort((a, b) => {
    const aIndex = languageOrder.indexOf(a);
    const bIndex = languageOrder.indexOf(b);

    if (aIndex > -1 && bIndex > -1) {
      return aIndex - bIndex;
    }

    if (aIndex > -1 && bIndex === -1) {
      return -1;
    }

    if (bIndex > -1 && aIndex === -1) {
      return 1;
    }

    return 0;
  });
}
