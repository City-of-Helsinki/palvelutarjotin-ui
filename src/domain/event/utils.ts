import { isToday, isTomorrow } from 'date-fns';
import { TFunction } from 'next-i18next';

import { EventFieldsFragment } from '../../generated/graphql';
import { Language } from '../../types';
import formatDate from '../../utils/formatDate';
import getLocalisedString from '../../utils/getLocalisedString';
import getTimeFormat from '../../utils/getTimeFormat';
import { EVENT_PLACEHOLDER_IMAGES } from './constants';

/**
 * Get event placeholder image url
 * @param {string} id
 * @return {string}
 */
export const getEventPlaceholderImage = (id: string): string => {
  const numbers = id.match(/\d+/g);
  const sum = numbers
    ? numbers.reduce((prev: number, cur: string) => prev + Number(cur), 0)
    : 0;
  const index = sum % 4;

  return EVENT_PLACEHOLDER_IMAGES[index];
};

export const getEventStartTimeStr = (
  event: EventFieldsFragment,
  locale: Language,
  t: TFunction
): string | null => {
  const startTime = event.startTime ? new Date(event.startTime) : null;
  const timeFormat = getTimeFormat(locale);

  if (!startTime) return null;

  if (isToday(startTime))
    return t('event:eventCard.startTime.today', {
      time: formatDate(startTime, timeFormat, locale),
    });

  if (isTomorrow(startTime))
    return t('event:eventCard.startTime.tomorrow', {
      time: formatDate(startTime, timeFormat, locale),
    });

  return t('event:eventCard.startTime.other', {
    date: formatDate(startTime),
    time: formatDate(startTime, timeFormat, locale),
  });
};

/**
 * Check is event free
 * @param eventData
 * @return {boolean}
 */
export const isEventFree = (event: EventFieldsFragment): boolean =>
  Boolean(event.offers.find((item) => item.isFree)?.isFree);

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const getEventFields = (
  event: EventFieldsFragment | undefined | null,
  locale: Language
) =>
  event
    ? {
        isEventFree: isEventFree(event),
        eventName: getLocalisedString(event?.name, locale),
        shortDescription: getLocalisedString(event?.shortDescription, locale),
        description: getLocalisedString(event?.description, locale),
        imageUrl: event?.images?.[0]?.url,
        imageAltText: event?.images?.[0]?.altText,
        photographerName: event?.images?.[0]?.photographerName,
        organization: event.pEvent?.organisation?.name,
        contactPhoneNumber: event.pEvent?.contactPhoneNumber,
        contactEmail: event.pEvent?.contactEmail,
        contactPerson: event.pEvent?.contactPerson?.name,
      }
    : {};
