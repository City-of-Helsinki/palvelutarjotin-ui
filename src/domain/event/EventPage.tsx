import take from 'lodash/take';
import { useRouter } from 'next/router';
import React, { ReactElement } from 'react';
import { useTranslation } from 'react-i18next';

import LoadingSpinner from '../../common/components/loadingSpinner/LoadingSpinner';
import { useEventQuery } from '../../generated/graphql';
import useLocale from '../../hooks/useLocale';
import Container from '../app/layout/Container';
import PageWrapper from '../app/layout/PageWrapper';
import { ROUTES } from '../app/routes/constants';
import NotFoundPage from '../notFoundPage/NotFoundPage';
import { OCCURRENCE_LIST_PAGE_SIZE } from './constants';
import EventBasicInfo from './eventBasicInfo/EventBasicInfo';
import EventImage from './eventImage/EventImage';
import styles from './eventPage.module.scss';
import Occurrences from './occurrences/Occurrences';
import { getEventFields } from './utils';

const EventPage = (): ReactElement => {
  const { t } = useTranslation();
  const locale = useLocale();
  const router = useRouter();
  const {
    query: { eventId },
  } = useRouter();
  const [occurrencesVisible, setOccurrencesVisible] = React.useState(
    OCCURRENCE_LIST_PAGE_SIZE
  );

  const { data: eventData, loading } = useEventQuery({
    variables: { id: eventId as string, include: ['keywords,location'] },
  });

  const enrolOccurrence = (occurrenceId: string) => {
    router.push({
      pathname: `/${locale}${ROUTES.CREATE_ENROLMENT.replace(
        ':id',
        eventId as string
      )}`,
      query: {
        occurrences: occurrenceId,
      },
    });
  };

  const showMoreOccurrences = () => {
    setOccurrencesVisible(occurrencesVisible + OCCURRENCE_LIST_PAGE_SIZE);
  };

  const {
    eventName,
    imageUrl,
    imageAltText,
    photographerName,
    occurrences,
    locationId,
    neededOccurrences,
  } = getEventFields(eventData?.event, locale);

  const visibleOccurrences = take(occurrences, occurrencesVisible);

  return (
    <PageWrapper title={eventName || t('event:pageTitle')}>
      <LoadingSpinner isLoading={loading}>
        {eventData?.event ? (
          <Container className={styles.eventPage}>
            <EventImage
              imageUrl={imageUrl}
              imageAltText={imageAltText}
              photographerName={photographerName}
            />
            <EventBasicInfo event={eventData.event} />
            {occurrences && (
              <Occurrences
                eventLocationId={locationId || ''}
                occurrences={visibleOccurrences}
                showMoreOccurrences={showMoreOccurrences}
                showMoreButtonVisible={occurrences.length > occurrencesVisible}
                enrolOccurrence={
                  neededOccurrences === 1 ? enrolOccurrence : null
                }
                // TODO: chooseOccurrence
              />
            )}
          </Container>
        ) : (
          <NotFoundPage />
        )}
      </LoadingSpinner>
    </PageWrapper>
  );
};

export default EventPage;
