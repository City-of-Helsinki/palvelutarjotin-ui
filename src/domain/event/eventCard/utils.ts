import { TFunction } from 'next-i18next';

import { EnrolmentStatusKeywordPropsType } from '../../../common/components/keyword/EnrolmentStatusKeyword';
import { EventsFieldsFragment } from '../../../generated/graphql';
import { Language } from '../../../types';
import { formatDateRange } from '../../../utils/time/format';
import { isMultidayOccurrence } from '../../occurrence/utils';
import { getEventStartTimeStr } from '../utils';

export const getOccurrenceDateStr = (
  occurrence: { startTime: string; endTime: string },
  locale: Language,
  t: TFunction
): string | null => {
  if (isMultidayOccurrence(occurrence)) {
    return formatDateRange(
      new Date(occurrence.startTime),
      new Date(occurrence.endTime)
    );
  }

  return getEventStartTimeStr(new Date(occurrence.startTime), locale, t);
};

/**
 * Has the event enrolments enabled in any of it's occurrences?
 * @param event The event to check the status for.
 * @returns `true` if any of the event occurrences has enrolments and
 *  `false` if none.
 */
export const hasEventEnrolmentsEnabled = (event: EventsFieldsFragment) =>
  Boolean(event.pEvent.enrolmentStart) || hasEventExternalEnrolments(event);

/**
 * Has the event external enrolments enabled in any of it's occurrences?
 *
 * @param event The event to check the status for.
 * @returns `true` if the event has some external enrolments and
 *  `false` if none of the occurrences has external enrolments.
 */
export const hasEventExternalEnrolments = (event: EventsFieldsFragment) =>
  Boolean(event.pEvent.externalEnrolmentUrl);

/**
 * Resolve the enrolment status of an event. Has the event free spots available?
 *
 * @param event The event to check the status for.
 * @returns Returns `null` if the enrolment status could not be resolved or
 *    the event has no enrolments at all,
 *    `true` if there are any free spots available and `false` if there are no
 *    spots available.
 */
export const hasEventFreeSpotsAvailable = (
  event: EventsFieldsFragment
): boolean | null => event.pEvent.hasSpaceForEnrolments ?? null;

/**
 * Determines the enrollment status of an event.
 *
 * @param event The event to check the enrollment status for.
 * @returns The enrollment status of the event, or `undefined`
 *      if no enrolments exist or the status could not be determined.
 */
export const getEventEnrolmentStatus = (
  event: EventsFieldsFragment
): EnrolmentStatusKeywordPropsType['enrolmentStatus'] => {
  if (hasEventEnrolmentsEnabled(event)) {
    if (hasEventExternalEnrolments(event)) {
      return 'externally_unknown';
    }
    const hasEventFreeSpots = hasEventFreeSpotsAvailable(event);

    if (hasEventFreeSpots === false) {
      // Event has no free spots available, so it is full.
      return 'full';
    } else if (hasEventFreeSpots === true) {
      // Event has some free spots available.
      return 'free_space_available';
    }
  }
  // No enrolments at all or the enrolment status could not be resolved
  return undefined;
};
