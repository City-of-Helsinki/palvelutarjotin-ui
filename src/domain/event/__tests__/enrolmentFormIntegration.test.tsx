/* eslint-disable @typescript-eslint/no-explicit-any */
import { MockedResponse } from '@apollo/client/testing';
import userEvent from '@testing-library/user-event';
import { advanceTo } from 'jest-date-mock';
import React from 'react';
import { act } from 'react-dom/test-utils';
import wait from 'waait';

import { FORM_NAMES } from '../../../constants';
import * as graphqlFns from '../../../generated/graphql';
import { createDefaultEventQueryMock } from '../../../tests/apollo-mocks/eventMocks';
import { createPlaceQueryMock } from '../../../tests/apollo-mocks/placeMocks';
import { createStudyLevelsQueryMock } from '../../../tests/apollo-mocks/studyLevelsMocks';
import { isFeatureEnabled } from '../../../utils/featureFlags';
import {
  fakeLocalizedObject,
  fakeOccurrences,
  fakePEvent,
  fakePlace,
} from '../../../utils/mockDataUtils';
import { getRecommendedEventsQueryVariables } from '../../../utils/recommendedEventsUtils';
import {
  render,
  configure,
  screen,
  waitFor,
  within,
  CustomRenderOptions,
} from '../../../utils/testUtils';
import { ROUTES } from '../../app/routes/constants';
import * as utils from '../../enrolment/utils';
import EventPage from '../EventPage';

configure({ defaultHidden: true });

// eslint-disable-next-line @typescript-eslint/no-explicit-any
(utils.getCAPTCHAToken as any) = jest.fn().mockResolvedValue('captcha-token');

const eventId = '123321';
const occurrenceId1 = '123321';
const occurrenceId2 = '321123';
const eventName = 'Testitapahtuma';
const locationId = 'locationId';
const locationName = 'Kirjasto';

const createPageMock = (
  pEventOverrides: Partial<graphqlFns.PalvelutarjotinEventNode>,
  locationId?: string
): MockedResponse[] => [
  createDefaultEventQueryMock({
    include: ['audience', 'in_language'],
    result: {
      id: eventId,
      name: fakeLocalizedObject(eventName),
      location: fakePlace(
        locationId
          ? { id: locationId, name: fakeLocalizedObject(locationName) }
          : {}
      ),
      pEvent: fakePEvent(pEventOverrides),
    },
  }),
  createStudyLevelsQueryMock(),
];

const occurrenceEnrolDifferentTimes: Partial<graphqlFns.OccurrenceNode>[] = [
  { id: occurrenceId1, startTime: new Date(2020, 8, 25, 12, 30) },
  { id: occurrenceId2, startTime: new Date(2020, 8, 26, 13, 20) },
  { id: 'sdfgdfhbf345345yreg', startTime: new Date(2020, 8, 26, 14, 20) },
];

const pageMockWithSelectedPlace: MockedResponse[] = [
  createPlaceQueryMock({
    id: locationId,
    name: fakeLocalizedObject(locationName),
  }),
  ...createPageMock(
    {
      enrolmentStart: new Date(2020, 8, 7).toISOString(),
      enrolmentEndDays: 3,
      occurrences: fakeOccurrences(3, occurrenceEnrolDifferentTimes),
      neededOccurrences: 2,
    },
    locationId
  ),
];

const pageMockEnrolHandpick: MockedResponse[] = createPageMock(
  {
    enrolmentStart: new Date(2020, 8, 7).toISOString(),
    enrolmentEndDays: 3,
    occurrences: fakeOccurrences(1, [
      { id: occurrenceId1, startTime: new Date(2020, 8, 25, 12, 30) },
    ]),
    neededOccurrences: 1,
    autoAcceptance: false,
    mandatoryAdditionalInformation: true,
  },
  locationId
);

const renderComponent = (options?: CustomRenderOptions) => {
  return render(<EventPage />, {
    query: { eventId: eventId },
    path: `/fi${ROUTES.EVENT_DETAILS.replace(':id', eventId)}`,
    ...options,
  });
};

advanceTo(new Date(2020, 8, 8));

// if these tests give bunch of act warnings that fill up console, comment test step out and add back one by one
// or the other way aroung + add "await act(() => wait(500))" as the last line in the test

test('user can select single occurrences and enrol to it with enrolment form', async () => {
  // eslint-disable-next-line import/namespace
  const enrolOccurrenceMock = jest.fn();
  jest
    .spyOn(graphqlFns, 'useEnrolOccurrenceMutation')
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    .mockReturnValue([enrolOccurrenceMock, { loading: false }] as any);

  renderComponent({
    mocks: pageMockEnrolHandpick,
  });

  const rows = await screen.findAllByRole('row');

  expect(rows[1].textContent).toMatchInlineSnapshot(
    `"25.9.2020 pe12:30 – 13:30suomi, englantifi, enKirjasto30 / 30Näytä lisätiedot"`
  );

  userEvent.click(
    within(rows[1]).getByRole('button', {
      name: /näytä tapahtuma-ajan lisätiedot/i,
    })
  );

  userEvent.click(screen.getByText(/varaustiedustelu/i));

  await screen.findByRole('heading', { name: 'Ilmoittajan tiedot' });

  await fillForm({
    additionalInfoName: /Lisätiedot \*/i,
    submitButtonName: /lähetä varaustiedustelu/i,
  });

  if (isFeatureEnabled('FORMIK_PERSIST')) {
    // wait for debounce to trigger and populate localStorage
    await act(() => wait(500));

    expect(
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      JSON.parse(localStorage.getItem(FORM_NAMES.ENROLMENT_FORM)!)
    ).toMatchSnapshot();
  }

  await waitFor(() => {
    expect(getRecommendedEventsQueryVariables()).toEqual({
      unitIds: [null],
      studyLevels: ['kultus:55', 'kultus:56'],
    });
  });

  await waitFor(() => {
    expect(enrolOccurrenceMock).toHaveBeenCalledWith({
      variables: {
        input: {
          captchaKey: 'captcha-token',
          notificationType: 'EMAIL_SMS',
          occurrenceIds: ['123321'],
          person: undefined,
          studyGroup: {
            amountOfAdult: 2,
            extraNeeds: 'Lisätietoja ilmoittautumiseen',
            groupName: '4a',
            groupSize: 10,
            unitId: null,
            unitName: 'Testikoulu',
            person: {
              emailAddress: 'testi@testi.fi',
              language: 'FI',
              name: 'Nimi Niminen',
              phoneNumber: '123321123',
            },
            studyLevels: ['GRADE_2', 'GRADE_4'],
          },
        },
      },
    });
  });
});

test('user can select multiple occurrences and enrol to them with enrolment form', async () => {
  // eslint-disable-next-line import/namespace
  const enrolOccurrenceMock = jest.fn();
  jest
    .spyOn(graphqlFns, 'useEnrolOccurrenceMutation')
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    .mockReturnValue([enrolOccurrenceMock, { loading: false }] as any);

  renderComponent({
    mocks: pageMockWithSelectedPlace,
  });

  await screen.findAllByRole('button', {
    name: /näytä tapahtuma-ajan lisätiedot/i,
  });

  const rows = await screen.findAllByRole('row');

  expect(rows[1].textContent).toMatchInlineSnapshot(
    `"25.9.2020 pe12:30 – 13:30suomi, englantifi, enKirjasto30 / 30Näytä lisätiedot"`
  );
  expect(rows[2].textContent).toMatchInlineSnapshot(
    `"26.9.2020 la13:20 – 14:20suomi, englantifi, enKirjasto30 / 30Näytä lisätiedot"`
  );
  expect(rows[3].textContent).toMatchInlineSnapshot(
    `"26.9.2020 la14:20 – 15:20suomi, englantifi, enKirjasto30 / 30Näytä lisätiedot"`
  );

  userEvent.click(
    within(rows[1]).getByRole('checkbox', { name: 'Valitse tapahtuma-aika' })
  );
  userEvent.click(
    within(rows[2]).getByRole('checkbox', { name: 'Valitse tapahtuma-aika' })
  );
  userEvent.click(screen.getByRole('button', { name: /ilmoittaudu/i }));

  await fillForm();

  if (isFeatureEnabled('FORMIK_PERSIST')) {
    // wait for debounce to trigger and populate localStorage
    await act(() => wait(500));

    expect(
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      JSON.parse(localStorage.getItem(FORM_NAMES.ENROLMENT_FORM)!)
    ).toMatchSnapshot();
  }

  await waitFor(() => {
    expect(enrolOccurrenceMock).toHaveBeenCalledWith({
      variables: {
        input: {
          captchaKey: 'captcha-token',
          notificationType: 'EMAIL_SMS',
          occurrenceIds: ['123321', '321123'],
          person: undefined,
          studyGroup: {
            amountOfAdult: 2,
            extraNeeds: 'Lisätietoja ilmoittautumiseen',
            groupName: '4a',
            groupSize: 10,
            unitId: null,
            unitName: 'Testikoulu',
            person: {
              emailAddress: 'testi@testi.fi',
              language: 'FI',
              name: 'Nimi Niminen',
              phoneNumber: '123321123',
            },
            studyLevels: ['GRADE_2', 'GRADE_4'],
          },
        },
      },
    });
  });
});

const fillForm = async ({
  additionalInfoName = /lisätiedot \(valinnainen\)/i,
  submitButtonName = /lähetä ilmoittautuminen/i,
}: {
  additionalInfoName?: RegExp | string;
  submitButtonName?: RegExp | string;
} = {}) => {
  const enrolmentForm = within(screen.getByTestId('enrolment-form'));

  userEvent.type(enrolmentForm.getByLabelText(/nimi/i), 'Nimi Niminen');

  userEvent.type(
    enrolmentForm.getByLabelText(/sähköpostiosoite/i),
    'testi@testi.fi'
  );

  userEvent.type(enrolmentForm.getByLabelText(/puhelinnumero/i), '123321123');

  userEvent.click(enrolmentForm.getByLabelText(/paikka ei ole listalla/i));

  userEvent.type(
    enrolmentForm.getByLabelText(/päiväkoti \/ koulu \/ oppilaitos/i),
    'Testikoulu'
  );
  userEvent.type(enrolmentForm.getByLabelText(/ryhmä/i), '4a');

  // select grade from dropdown
  userEvent.click(enrolmentForm.getByRole('button', { name: /luokka-aste/i }));

  const grade1 = await enrolmentForm.findByRole('option', {
    name: /2\. luokka/i,
  });
  const grade2 = await enrolmentForm.findByRole('option', {
    name: /4\. luokka/i,
  });
  userEvent.click(grade1);
  userEvent.click(grade2);

  // close dropdown
  userEvent.click(enrolmentForm.getByRole('button', { name: /luokka-aste/i }));

  userEvent.type(enrolmentForm.getByLabelText(/lapsia/i), '10');
  userEvent.type(enrolmentForm.getByLabelText(/aikuisia/i), '2');

  expect(
    enrolmentForm.getByRole('checkbox', { name: /sama kuin ilmoittaja/i })
  ).toBeChecked();
  userEvent.click(
    enrolmentForm.getByRole('checkbox', { name: /sähköpostilla/i })
  );
  userEvent.click(enrolmentForm.getByText(/tekstiviestillä/i));

  userEvent.click(
    enrolmentForm.getByRole('button', { name: /Ilmoitusten kieli/ })
  );
  userEvent.click(enrolmentForm.getByRole('option', { name: /suomi/i }));

  userEvent.type(
    enrolmentForm.getByRole('textbox', { name: additionalInfoName }),
    'Lisätietoja ilmoittautumiseen'
  );

  // test that error notification works
  userEvent.click(
    enrolmentForm.getByRole('button', { name: submitButtonName })
  );
  const alertContainer = await enrolmentForm.findByRole('alert');
  expect(alertContainer).toHaveTextContent(/Virhe lomakkeessa/i);
  expect(alertContainer).toHaveTextContent(
    /hyväksyn tietojeni jakamisen tapahtuman järjestäjän kanssa/i
  );

  userEvent.click(
    enrolmentForm.getByRole('checkbox', {
      name: /hyväksyn tietojeni jakamisen tapahtuman järjestäjän kanssa/i,
    })
  );

  await waitFor(() => {
    expect(enrolmentForm.queryByRole('alert')).not.toBeInTheDocument();
  });

  userEvent.click(
    enrolmentForm.getByRole('button', { name: submitButtonName })
  );
};
