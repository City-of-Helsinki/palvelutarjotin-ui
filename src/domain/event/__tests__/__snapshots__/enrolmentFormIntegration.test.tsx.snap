// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user can select multiple occurrences and enrol to them with enrolment form 4`] = `
Object {
  "dirty": true,
  "errors": Object {},
  "initialErrors": Object {},
  "initialTouched": Object {},
  "initialValues": Object {
    "hasEmailNotification": true,
    "hasSmsNotification": false,
    "isMandatoryAdditionalInformationRequired": false,
    "isSameResponsiblePerson": true,
    "isSharingDataAccepted": false,
    "language": "FI",
    "person": Object {
      "emailAddress": "",
      "name": "",
      "phoneNumber": "",
    },
    "studyGroup": Object {
      "amountOfAdult": "",
      "extraNeeds": "",
      "groupName": "",
      "groupSize": "",
      "person": Object {
        "emailAddress": "",
        "name": "",
        "phoneNumber": "",
      },
      "studyLevels": Array [],
      "unitId": "",
      "unitName": "",
    },
  },
  "isSubmitting": false,
  "isValid": true,
  "isValidating": false,
  "submitCount": 2,
  "touched": Object {
    "hasEmailNotification": true,
    "hasSmsNotification": true,
    "isMandatoryAdditionalInformationRequired": true,
    "isSameResponsiblePerson": true,
    "isSharingDataAccepted": true,
    "language": true,
    "person": Object {
      "emailAddress": true,
      "name": true,
      "phoneNumber": true,
    },
    "studyGroup": Object {
      "amountOfAdult": true,
      "extraNeeds": true,
      "groupName": true,
      "groupSize": true,
      "person": Object {
        "emailAddress": true,
        "name": true,
        "phoneNumber": true,
      },
      "studyLevels": Array [
        true,
        true,
      ],
      "unitId": true,
      "unitName": true,
    },
  },
  "validateOnBlur": true,
  "validateOnChange": true,
  "validateOnMount": false,
  "values": Object {
    "hasEmailNotification": true,
    "hasSmsNotification": true,
    "isMandatoryAdditionalInformationRequired": false,
    "isSameResponsiblePerson": true,
    "isSharingDataAccepted": true,
    "language": "FI",
    "person": Object {
      "emailAddress": "",
      "name": "",
      "phoneNumber": "",
    },
    "studyGroup": Object {
      "amountOfAdult": 2,
      "extraNeeds": "Lisätietoja ilmoittautumiseen",
      "groupName": "4a",
      "groupSize": 10,
      "person": Object {
        "emailAddress": "testi@testi.fi",
        "name": "Nimi Niminen",
        "phoneNumber": "123321123",
      },
      "studyLevels": Array [
        "GRADE_2",
        "GRADE_4",
      ],
      "unitId": null,
      "unitName": "Testikoulu",
    },
  },
}
`;

exports[`user can select multiple occurrences and enrol to them with enrolment form 5`] = `
Object {
  "dirty": true,
  "errors": Object {},
  "initialErrors": Object {},
  "initialTouched": Object {},
  "initialValues": Object {
    "hasEmailNotification": true,
    "hasSmsNotification": false,
    "isMandatoryAdditionalInformationRequired": false,
    "isSameResponsiblePerson": true,
    "isSharingDataAccepted": false,
    "language": "FI",
    "person": Object {
      "emailAddress": "",
      "name": "",
      "phoneNumber": "",
    },
    "studyGroup": Object {
      "amountOfAdult": "",
      "extraNeeds": "",
      "groupName": "",
      "groupSize": "",
      "person": Object {
        "emailAddress": "",
        "name": "",
        "phoneNumber": "",
      },
      "studyLevels": Array [],
      "unitId": "",
      "unitName": "",
    },
  },
  "isSubmitting": false,
  "isValid": true,
  "isValidating": false,
  "submitCount": 2,
  "touched": Object {
    "hasEmailNotification": true,
    "hasSmsNotification": true,
    "isMandatoryAdditionalInformationRequired": true,
    "isSameResponsiblePerson": true,
    "isSharingDataAccepted": true,
    "language": true,
    "person": Object {
      "emailAddress": true,
      "name": true,
      "phoneNumber": true,
    },
    "studyGroup": Object {
      "amountOfAdult": true,
      "extraNeeds": true,
      "groupName": true,
      "groupSize": true,
      "person": Object {
        "emailAddress": true,
        "name": true,
        "phoneNumber": true,
      },
      "studyLevels": true,
      "unitId": true,
      "unitName": true,
    },
  },
  "validateOnBlur": true,
  "validateOnChange": true,
  "validateOnMount": false,
  "values": Object {
    "hasEmailNotification": true,
    "hasSmsNotification": true,
    "isMandatoryAdditionalInformationRequired": false,
    "isSameResponsiblePerson": true,
    "isSharingDataAccepted": true,
    "language": "FI",
    "person": Object {
      "emailAddress": "",
      "name": "",
      "phoneNumber": "",
    },
    "studyGroup": Object {
      "amountOfAdult": 2,
      "extraNeeds": "Lisätietoja ilmoittautumiseen",
      "groupName": "4a",
      "groupSize": 10,
      "person": Object {
        "emailAddress": "testi@testi.fi",
        "name": "Nimi Niminen",
        "phoneNumber": "123321123",
      },
      "studyLevels": Array [
        "GRADE_2",
        "GRADE_4",
      ],
      "unitId": null,
      "unitName": "Testikoulu",
    },
  },
}
`;

exports[`user can select single occurrences and enrol to it with enrolment form 2`] = `
Object {
  "dirty": true,
  "errors": Object {},
  "initialErrors": Object {},
  "initialTouched": Object {},
  "initialValues": Object {
    "hasEmailNotification": true,
    "hasSmsNotification": false,
    "isMandatoryAdditionalInformationRequired": true,
    "isSameResponsiblePerson": true,
    "isSharingDataAccepted": false,
    "language": "FI",
    "person": Object {
      "emailAddress": "",
      "name": "",
      "phoneNumber": "",
    },
    "studyGroup": Object {
      "amountOfAdult": "",
      "extraNeeds": "",
      "groupName": "",
      "groupSize": "",
      "person": Object {
        "emailAddress": "",
        "name": "",
        "phoneNumber": "",
      },
      "studyLevels": Array [],
      "unitId": "",
      "unitName": "",
    },
  },
  "isSubmitting": false,
  "isValid": true,
  "isValidating": false,
  "submitCount": 2,
  "touched": Object {
    "hasEmailNotification": true,
    "hasSmsNotification": true,
    "isMandatoryAdditionalInformationRequired": true,
    "isSameResponsiblePerson": true,
    "isSharingDataAccepted": true,
    "language": true,
    "person": Object {
      "emailAddress": true,
      "name": true,
      "phoneNumber": true,
    },
    "studyGroup": Object {
      "amountOfAdult": true,
      "extraNeeds": true,
      "groupName": true,
      "groupSize": true,
      "person": Object {
        "emailAddress": true,
        "name": true,
        "phoneNumber": true,
      },
      "studyLevels": Array [
        true,
        true,
      ],
      "unitId": true,
      "unitName": true,
    },
  },
  "validateOnBlur": true,
  "validateOnChange": true,
  "validateOnMount": false,
  "values": Object {
    "hasEmailNotification": true,
    "hasSmsNotification": true,
    "isMandatoryAdditionalInformationRequired": true,
    "isSameResponsiblePerson": true,
    "isSharingDataAccepted": true,
    "language": "FI",
    "person": Object {
      "emailAddress": "",
      "name": "",
      "phoneNumber": "",
    },
    "studyGroup": Object {
      "amountOfAdult": 2,
      "extraNeeds": "Lisätietoja ilmoittautumiseen",
      "groupName": "4a",
      "groupSize": 10,
      "person": Object {
        "emailAddress": "testi@testi.fi",
        "name": "Nimi Niminen",
        "phoneNumber": "123321123",
      },
      "studyLevels": Array [
        "GRADE_2",
        "GRADE_4",
      ],
      "unitId": null,
      "unitName": "Testikoulu",
    },
  },
}
`;
